File: backend/app/flow_v2/data_collection_handles.py
"""Canonical edge-handle contract for Data Collection nodes."""
from __future__ import annotations

from typing import Any

SUCCESS = "success"
INVALID = "invalid"
CANCEL = "cancel"
TIMEOUT = "timeout"
HANDLES = frozenset({SUCCESS, INVALID, CANCEL, TIMEOUT})
LEGACY_ALIASES = {"retry_exhausted": INVALID}


def normalize_data_collection_handle(value: Any) -> str:
    handle = str(value or "").strip().lower()
    return LEGACY_ALIASES.get(handle, handle)


def normalize_data_collection_edges(nodes: list[dict[str, Any]], edges: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Migrate explicit legacy handles without guessing a missing connection."""
    collection_ids = {
        str(node.get("id")) for node in nodes
        if isinstance(node, dict) and str(node.get("type") or (node.get("data") or {}).get("type") or "").lower() == "data_collection"
    }
    normalized: list[dict[str, Any]] = []
    for edge in edges:
        if not isinstance(edge, dict) or str(edge.get("source")) not in collection_ids:
            normalized.append(edge)
            continue
        data = edge.get("data") if isinstance(edge.get("data"), dict) else {}
        raw = edge.get("sourceHandle", edge.get("source_handle"))
        if raw is None:
            raw = data.get("sourceHandle", data.get("source_handle"))
        handle = normalize_data_collection_handle(raw)
        if not handle:
            normalized.append(edge)
            continue
        normalized.append({**edge, "sourceHandle": handle, "data": {**data, "sourceHandle": handle}})
    return normalized

File: frontend/lib/dataCollectionHandles.ts
import type { Edge, Node } from 'reactflow';

export const DATA_COLLECTION_HANDLES = ['success', 'invalid', 'cancel', 'timeout'] as const;
export type DataCollectionHandle = typeof DATA_COLLECTION_HANDLES[number];

const LEGACY_ALIASES: Record<string, DataCollectionHandle> = { retry_exhausted: 'invalid' };

export const normalizeDataCollectionHandle = (value: unknown): string => {
  const handle = String(value ?? '').trim().toLowerCase();
  return LEGACY_ALIASES[handle] || handle;
};

/** Migrate explicit legacy IDs only; never guess/reconnect an absent handle. */
export const normalizeDataCollectionEdges = <T extends Edge>(nodes: Node[], edges: T[]): T[] => {
  const collectionIds = new Set(nodes.filter((node) => node.type === 'data_collection').map((node) => node.id));
  return edges.map((edge) => {
    if (!collectionIds.has(edge.source)) return edge;
    const raw = edge.sourceHandle ?? edge.data?.sourceHandle ?? edge.data?.source_handle;
    const sourceHandle = normalizeDataCollectionHandle(raw);
    if (!sourceHandle) return edge;
    return { ...edge, sourceHandle, data: { ...(edge.data || {}), sourceHandle } };
  });
};
